txpipe/twopoint.py

```diff
@@ -202,9 +221,118 @@ def _read_nbin_from_config(self):
+        bz={}
+        fidCosmo=ccl.Cosmology(**fid_params)
+        for key,tracer in sacc.tracers.items():
+            if 'lens' in key:
+                zeff = (tracer.z*tracer.nz).sum()/tracer.nz.sum()
+                bz[key] = self.config['blind_b0']/ccl.growth_factor(fidCosmo,1/(1+zeff)) 
+        
+        offset_params = copy.copy(fid_params)
+        for par in fid_params.keys():
+            offset_params [par] += self.config['blind_'+par][1]*np.random.normal(0.,1.)
+        fc_config = """
+           parameters:
+             ## missing params set in fid_params and offset_params
+             Omega_k: 0.0
+             wa: 0.0
+
+             one: 1
+"""
+        for k,v in bz.items():
+            fc_config+="             bias_{lens}: {bias}\n".format(lens=k,bias=v)
+        fc_config += """
+           two_point:
+             module: firecrown.ccl.two_point
+           
+             systematics:
+               dummy:
+                  kind: PhotoZShiftBias
+                  delta_z: one
+
+        """
+
+        ### sources
+        fc_config+="     sources: \n"
+        srclist=[]
+        lenslist=[]
+        for key,tracer in sacc.tracers.items():
+            ## This is a hack, need to think how to do this better
+            if 'source' in key:
+                fc_config+="""
+                                  {src}:
+                                     kind: WLSource
+                                     sacc_tracer: {src}\n""".format(src=key)
+                srclist.append(key)
+            if 'lens' in key:
+                fc_config+="""
+                                  {lens}:
+                                     kind: NumberCountsSource
+                                     bias: bias_{lens}
+                                     sacc_tracer: {lens}\n""".format(lens=key)
+                lenslist.append(key)
+        fc_config+="             statistics: \n"
+        ##
+        ## list(dict.fromkeys()) gives unique elements while preserving order
+        types=list(dict.fromkeys([(p.data_type, p.tracers,
+                                   "{dtype}_{tracer1}_{tracer2}".format(dtype=p.data_type,
+                                            tracer1=p.tracers[0],tracer2=p.tracers[1]))
+                                   for p in sacc.data]))
+        for dtype,(tracer1,tracer2),fcname in types:
+            fc_config+="""
+                                 {fcname}:
+                                   sources: ['{tracer1}', '{tracer2}']
+                                   sacc_data_type: {dtype}
+                   """.format(fcname=fcname, dtype=dtype,tracer1=tracer1,tracer2=tracer2)
+        ## now try to get predictions
+        pred={}
+        for name,pars in [('fid',fid_params),('ofs',offset_params)]:
+            print ("Calling firecrown : %s"%(name))
+
+            config, data = firecrown.parse(io.StringIO(fc_config),
+               settings={'two_point':{'sacc_data':sacc},'parameters':pars})
```

Comment:
So @slosar simply make the edits you'd like to config and data here in the code. And then feed those to firecrown. This will keep the code I don't like out of firecrown.

Reply:
Ok, I can pass cosmological parameters as strings rather than floats (it is ugly in my view, but ok), but I cannot pass a sacc object as a string. The only way to do it is dump it to disc, which is to start with very inefficient at nersc (takes minutes to dump a 2Mb sacc file), but important it is just wrong to be interacting with a library via objects on disk.

Reply:
Right so sacc object is def on the firecrown side. I will push a PR.